src/web/lib/braze/LocalMessageCache.ts

```diff
@@ -0,0 +1,76 @@
+import type appboy from '@braze/web-sdk-core';
+import { SlotName } from './BrazeMessages';
+
+const localStorageKeyBase = 'gu.brazeMessageCache';
+
+type Message = appboy.InAppMessage;
+
+interface Store {
+	getItem: (key: string) => string | null;
+	setItem: (key: string, value: string) => void;
+	removeItem: (key: string) => void;
+}
+
+const keyFromSlotName = (slotName: SlotName): string =>
+	`${localStorageKeyBase}.${slotName}`;
+
+class LocalMessageCache {
+	store: Store;
+
+	constructor(store: Store) {
+		this.store = store;
+	}
+
+	shift(slotName: SlotName): Message | undefined {
+		const key = keyFromSlotName(slotName);
+		const queue = this.readQueue(key);
+		const topItem = queue.shift();
+
+		if (topItem) {
+			this.store.setItem(key, JSON.stringify(queue));
+			return topItem;
+		}
+	}
+
+	peek(slotName: SlotName): Message | undefined {
```

Comment:
What's the motivation for `peek` returning the actual message? It seems to me that this interface makes it easy for clients to misuse this interface.

Peek allows client code to get the entire message and display it, without removing the message from the store. If peek instead only said that a message exists (`bool` - all ew technically need), the number of messages (`number` - probably the best balance of usability and strictness) or perhaps even just a summary of the messages (`array<string>` - I'm not a fan of this), then we'd be addressing the use case of "is there a message for me" while strictly enforcing the correct usage of the cache.

Reply:
Really good point, I agree that would be much better. In that case this could become `hasMessage(slotName: SlotName): boolean` which seems nicer. I can't remember if I did things this way because it was necessary because of how the rest of the message selection stuff is wired up, I'll investigate.

Reply:
Would this complicate e.g. the Epic where we don't want to remove message from the queue until it's actually been seen? Do we want to decouple using a message from deciding it has been displayed - otherwise we will have to handle re-adding an undisplayed Epic to the front of the queue.

Reply:
Thanks @rhystmills this is a really good point. Yeah for the epic we probably want to render it into the DOM but not remove it from the queue until it's had an impression. The alternative would be to pop the message off and then push back on if it doesn't get seen, but it's tricky to know when that would happen (when the page gets unloaded?)